Site list v2: retain `fields` and type-specific `layout` when changing view type

fushar opened the pull request:
Fixes DOTDASH-122
Fixes DOTDASH-127

## Proposed Changes

When changing v2 site list view to another type, before this PR:

- We could "pollute" the preferences with invalid fields from the previous type. For example, when we hide site icon from `table` view, `showMedia: false` will be persisted, but then when we change the type to `grid`, we lose the preview but we cannot get it back because that field is unmodifiable via https://github.com/Automattic/wp-calypso/pull/104628.
- We lose type-specific layout appearance options, such as density (table view) or preview size (grid view).

With this PR, we:

- Restore the type's default `showMedia` and type-specific `layout`.
- If the user has not explicitly modified the default fields, we will retain the fields, otherwise we restore the type's default fields.

## Why are these changes being made?

Fixes CfT bugs.

## Testing Instructions

1. It's best to first clear the `sites-view` Calypso preferences first via dev console's `POST /me/preferences`.
1. Go to /sites.
2. Verify you see the table view by default (if you have more than 10 sites... obviously you do)
3. Change to to grid view.
    - Verify the transition is smooth
    - Verify you see the default grid view fields.
4. Change back to table view.
    - Verify the transition is smooth
    - Verify you see the default table view fields.
5. Verify the logic in the PR's description.
6. Try playing around with appearance settings; verify they are persisted as you would expect from the UI.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [x] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

fushar committed:
Site list v2: reset view back to default when changing layout type

p-jackson commented:
I guess resetting them all is better than resetting only some. But wouldn't the other fix—not resetting any settings—be better? If the user has indicated that they care about the "outdated plugins" field, so they've enabled it, why wouldn't we take that signal and keep that field shown in the other table mode? Same as sorting. You've gotten the sites to the top of the list you want, now you change modes so you can see the preview, and boom it has disappeared.

If we do think of grid and table as fundamentally different views, then perhaps another approach would be to store the preferences under different keys. So when you switch to grid view we show you the last grid view preferences you had.

fushar commented:
Thanks for the comment, let me rethink the approach.

arthur791004 commented:
I guess we can simply reset `showMedia` to true when changing layout type 🤔

fushar committed:
Persist type-specific preferences

fushar commented:
PLEASE don't try to refresh the page while it has query params. Somehow, we have this issue where the query params are not deserialized to `View` on page refresh, even on trunk 🤔 😭 in only works on a single navigation session without refreshes...

<img width="870" alt="image" src="https://github.com/user-attachments/assets/2ad33de5-12e5-4751-90d1-05677f9066f7" />


<img width="983" alt="image" src="https://github.com/user-attachments/assets/72d582b0-2e09-4d5b-a5ea-52761c1b2ce4" />

I'll try to investigate this as separate issue.

arthur791004 commented:
It became somewhat more complex than I was aiming to avoid 🙈

fushar committed:
Just merge the view layout fields

fushar commented:
I've simplified the logic a lot, and updated the PR description. I think this is the best I can do to improve the current behavior. Ready for another review 😄

fushar commented on client/dashboard/sites/views.ts:
```diff
@@ -1,50 +1,70 @@
 
-export const PERSISTABLE_VIEW_KEYS = [
-	'density',
+const DEFAULT_LAYOUT_FIELDS: SupportedLayouts = {
```
We need to extract this outside `DEFAULT_LAYOUTS` (which is passed to `DataViews`), otherwise we will always see a flash of default fields every time we change the type.

fushar commented on client/dashboard/sites/views.ts:
```diff
@@ -1,50 +1,70 @@
 };
 
-export const PERSISTABLE_VIEW_KEYS = [
-	'density',
```
`density` is nested inside `layout`; it's not a top-level field.

fushar commented on client/dashboard/sites/views.ts:
```diff
@@ -1,50 +1,70 @@
 	'fields',
 	'layout',
 	'perPage',
-	'previewSize',
```
`previewSize` is nested inside `layout`; it's not a top-level field.

fushar commented on client/dashboard/sites/views.ts:
```diff
@@ -80,31 +101,118 @@ export function getView( {
+		// Show only params which have custom values.
+		...pickNonDefaultFields( updatedView, VIEW_SEARCH_PARAM_KEYS, defaultNextView ),
+
+		// Show the type param explicitly to ensure the view type is updated immediately.
```
This is what fixes DOTDASH-127.

fushar commented on client/dashboard/app/queries/me-preferences.ts:
```diff
@@ -7,7 +7,7 @@ export const userPreferencesQuery = ( preferenceName?: keyof UserPreferences ) =
 	select: ( data: UserPreferences ) => {
 		if ( preferenceName ) {
-			return data[ preferenceName ];
+			return data[ preferenceName ] || {};
```
I sometimes got "WSOD" (in a Suspense) without this... not sure how it ever worked before 🤔

p-jackson replied:
Oh dear, this is actually a bit of a problem. ReactQuery thinks that `undefined` means that the queries data hasn't been loaded yet. However for us `undefined` is a totally valid state (the user hasn't saved any view preferences yet).

Not only that, but you're initialising a user preference to some default value (i.e. `{}`) if it has never been saved before. I believe Calypso v1 does the same thing, there's a big table of default preferences.

The problem is that `{}` is only appropriate for this one preference we have. In the future we'll have preferences that are strings and can't be initialised this way.

I don't think it blocks this PR, but I'm working on some TypeScript shenanigans locally that I'll push separately.

fushar replied:
Updated so that the caller provides the default value: 6cd9d57cdea

It's not the best solution, didn't have time to fight TypeScript types 😄 but should work for now 😄

p-jackson commented on client/dashboard/sites/index.tsx:
```diff
@@ -76,33 +77,21 @@ export default function Sites() {
-			...defaultView,
-			...DEFAULT_LAYOUTS[ nextView.type as keyof typeof DEFAULT_LAYOUTS ],
-		};
+		const { updatedViewPreferences, updatedViewSearchParams } = getUpdatedView( {
```
`get` was a little confusing to me. Maybe something like `mergeViewPreferences`? Or `mergeViews`?

fushar replied:
Renamed to `mergeViews` and added comment in f76f5e4b9ca.

fushar committed:
Caller provides preferences default value

fushar committed:
Rename getUpdatedView -> mergeViews and add comment

arthur791004 commented on client/dashboard/app/queries/me-preferences.ts:
```diff
@@ -2,12 +2,15 @@ import { fetchPreferences, updatePreferences } from '../../data/me-preferences';
 	select: ( data: UserPreferences ) => {
 		if ( preferenceName ) {
-			return data[ preferenceName ];
+			return data[ preferenceName ] || defaultValue;
```
Somehow, we need to ensure that a default value is required whenever a preferenceName is provided. Otherwise, the same issue will occur 🤔

fushar replied:
Yes, but TS won't allow me to have required param (`defaultValue`) after an optional one (`preferenceName`). That's why I said it's not the best solution 😄 

I think we need to split the query into two (all data vs one preference), or come up with a better TypeScript typing. But I don't think it should block this PR; it's an existing issue.

p-jackson replied:
Yeah, lets fix the TS in another PR

arthur791004 commented on client/dashboard/sites/views.ts:
```diff
@@ -80,31 +101,122 @@ export function getView( {
+
+				// If the type is changed, and the user has never explicitly set custom fields,
+				// we restore the default fields for that type.
+				...DEFAULT_LAYOUT_FIELDS[ nextType ],
```
Looks like we can keep it in `DEFAULT_LAYOUTS` and then override the fields if the user has explicitly set custom fields 🤔

p-jackson replied:
I think it's a more general problem than just for the site layout preference. I think we need to provide a default value for _all_ user preferences. We can't just let the query return `undefined` because TanStack then can't tell the difference between the data not being available—and `useSuspenseQuery` throws—and that the preference just hasn't be saved yet.

fushar replied:
> Looks like we can keep it in DEFAULT_LAYOUTS

No, we can't. It is passed as `defaultLayouts` prop to `DataViews`, which will flash those default fields initially.

In Core, we don't pass the default fields to that prop. See e.g.: 

https://github.com/WordPress/gutenberg/blob/5d2f86a1e0db8708d875b1ca04f9e25541a822b0/packages/edit-site/src/components/page-templates/index.js#L35-L52

https://github.com/WordPress/gutenberg/blob/5d2f86a1e0db8708d875b1ca04f9e25541a822b0/packages/edit-site/src/components/page-templates/index.js#L66

fushar replied:
This is the flash that happens if we keep them in `DEFAULT_LAYOUTS`:


https://github.com/user-attachments/assets/8c7bb48c-f209-4546-a287-3e35d26aa612

p-jackson reviewed (approved):
😮‍💨 phew it feels like so much code to do a standard thing. I guess it is just the nature of the dataview's API, that it funnels everything through one uber `onViewChange` event.

A quicker way to clear the `sites-view` preference is here.

![CleanShot 2025-07-10 at 13 57 48@2x](https://github.com/user-attachments/assets/e4a40848-8d76-45bf-af6e-e1b9083042d7)

fushar commented:
Shipping, thanks all for the review!
